### PR TITLE
Searchbar

### PR DESCRIPTION
**Pull request by hernan-clich**
## Summary

- Added a magnifying glass icon in the `Library` screen that redirects to the new `Search` screen.
- If the input field is empty, the `Search` screen will display the placeholder screen.
- If the input field is not empty it will either filter if there are books that match that text, or display a disclaimer message stating that No books were found.
- User searches are case insensitive.
- The user can clear the search input by clicking the right side (X) icon.
- After finding the searched book, the user can access its corresponding `Book details` page.
- The user can then go back to the `Search` screen when clicking the back button.

#### Galaxy S (or similar)

rn_searchbar_galaxyS

#### Pixel 2/3 (or similar)

rn_searchbar_pixel3

#### Pixel 3XL (or similar)

rn_searchbar_pixel3xl

## Trello Card

https://trello.com/c/lErbUI6T/31-n-112-filtro-de-b%C3%BAsqueda

### CODE REVIEW

File: wbooks/src/app/components/AppNavigator/config.tsx [hunk 1]
```diff
@@ -2,20 +2,31 @@ import React from 'react';
...
 import styles from './styles';
 
-const customHeader = {
+const sharedHeaderProperties = {
   headerStyle: {
     height: 90
   },
-  headerTintColor: colors.white,
-  headerBackImage: () => <Image style={styles.backButton} resizeMode="cover" source={backButton} />,
   headerBackground: () => <Image style={styles.headerImage} resizeMode="stretch" source={headerBackground} />,
```
Comment (sfernandez11) [inline wbooks/src/app/components/AppNavigator/config.tsx]: Why these changes?

File: wbooks/src/app/components/AppNavigator/config.tsx [hunk 2]
```diff
@@ -2,20 +2,31 @@ import React from 'react';
...
     textTransform: 'uppercase'
   }
 };
 
-export default customHeader;
+const mainHeader = {
+  ...sharedHeaderProperties,
+  headerTintColor: colors.white,
+  headerBackImage: () => <Image style={styles.backButton} resizeMode="cover" source={backButton} />
+};
```
Comment (sfernandez11) [inline wbooks/src/app/components/AppNavigator/config.tsx]: Why? The `headerBackImage`  must be in all the `Stack` screens

File: wbooks/src/app/components/AppNavigator/config.tsx [hunk 3]
```diff
@@ -2,20 +2,31 @@ import React from 'react';
...
+
+export const searchHeader = {
+  ...sharedHeaderProperties,
```
Comment (sfernandez11) [inline wbooks/src/app/components/AppNavigator/config.tsx]: Not necessary, the Stack send the base config to the screens

File: wbooks/src/app/components/AppNavigator/index.tsx [hunk 4]
```diff
@@ -1,32 +1,39 @@
...
 
 function LibraryStackScreen() {
   return (
-    <LibraryStackNavigator.Navigator screenOptions={customHeader as object}>
+    <LibraryStackNavigator.Navigator screenOptions={mainHeader as object}>
       <LibraryStackNavigator.Screen
         name={SCREENS.LIBRARY}
         component={Library}
-        options={{ title: SCREEN_TITLES.LIBRARY }}
+        options={{ title: SCREEN_TITLES.LIBRARY, headerRight: () => <SearchIcon /> }}
```
Comment (sfernandez11) [inline wbooks/src/app/components/AppNavigator/index.tsx]: extract this `options` too

File: wbooks/src/app/components/AppNavigator/config.tsx [hunk 5]
```diff
@@ -2,20 +2,31 @@ import React from 'react';
...
+  headerLeft: () => null,
```
Comment (sfernandez11) [inline wbooks/src/app/components/AppNavigator/config.tsx:30]: Why?

File: wbooks/src/app/components/AppNavigator/config.tsx [hunk 6]
```diff
@@ -2,20 +2,31 @@ import React from 'react';
...
+  headerTitle: () => <SearchBar />
+};
+
+export default mainHeader;
```
Comment (sfernandez11) [inline wbooks/src/app/components/AppNavigator/config.tsx]: Why? Use `export const` when you have more than one object in a file

File: wbooks/src/app/components/AppNavigator/index.tsx [hunk 7]
```diff
@@ -1,32 +1,39 @@
...
 import colors from '@constants/colors';
 import SCREENS, { SCREEN_TITLES } from '@constants/screens';
 import { BottomTabParamList, LibraryStackParamList } from '@interfaces/navigatorParamLists';
 import BookDetails from '@screens/BookDetails';
 import Library from '@screens/Library';
+import SearchScreen from '@screens/SearchScreen';
 import WishList from '@screens/WishList';
 
-import customHeader from './config';
+import mainHeader, { searchHeader } from './config';
```
Comment (sfernandez11) [inline wbooks/src/app/components/AppNavigator/index.tsx]: Wrong!

File: wbooks/src/app/components/AppNavigator/index.tsx [hunk 8]
```diff
@@ -1,32 +1,39 @@
...
 
 const TabNavigator = createBottomTabNavigator<BottomTabParamList>();
 const LibraryStackNavigator = createStackNavigator<LibraryStackParamList>();
 
 function LibraryStackScreen() {
   return (
-    <LibraryStackNavigator.Navigator screenOptions={customHeader as object}>
+    <LibraryStackNavigator.Navigator screenOptions={mainHeader as object}>
```
Comment (sfernandez11) [inline wbooks/src/app/components/AppNavigator/index.tsx:20]: Revert changes!

File: wbooks/src/app/components/SearchEmpty/index.tsx [hunk 9]
```diff
@@ -0,0 +1,17 @@
+import React from 'react';
+import { Image, Text, View } from 'react-native';
+import searchPlaceholderIcon from '@assets/ic_search_placeholder.png';
+
+import styles from './styles';
+
+function SearchScreen() {
```
Comment (sfernandez11) [inline wbooks/src/app/components/SearchEmpty/index.tsx]: Bad copy/paste

File: wbooks/src/app/components/SearchBar/styles.ts [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+  },
+  searchIconLeft: {
+    height: 12,
+    tintColor: colors.manatee,
+    width: 13
+  },
+  clearIconRight: {
+    height: 14,
+    tintColor: colors.manatee,
+    width: 14
```
Comment (sfernandez11) [inline wbooks/src/app/components/SearchBar/styles.ts]: Extract common size = 14

File: wbooks/src/app/components/SearchEmpty/styles.ts [hunk 11]
```diff
@@ -0,0 +1,24 @@
...
+  container: {
+    alignItems: 'center',
+    backgroundColor: colors.polar,
+    flex: 1,
+    justifyContent: 'center'
+  },
+  searchIcon: {
+    height: 50,
+    marginBottom: 20,
+    width: 50
```
Comment (sfernandez11) [inline wbooks/src/app/components/SearchEmpty/styles.ts]: Same, don't repeat => `ICON_SIZE = 50` for example

File: wbooks/src/app/screens/Library/index.tsx [hunk 12]
```diff
@@ -1,45 +1,24 @@
...
-      bookTitle={trimLineBreaks(item.title)}
-      bookAuthor={item.author}
-      onPressBook={() => navigation.navigate(SCREENS.BOOK_DETAILS)}
-      style={index === 0 ? styles.firstItem : {}}
-    />
-  );
-  const keyExtractor = ({ id }: Book) => String(id);
-
   return (
     <SafeAreaView style={styles.container}>
```
Comment (sfernandez11) [inline wbooks/src/app/screens/Library/index.tsx:12]: The `BookList` component already have the `SafeAreaView`

File: wbooks/src/app/screens/Library/index.tsx [hunk 13]
```diff
@@ -1,45 +1,24 @@
...
-import { Book } from '@interfaces/books';
+import BookList from '@components/BookList';
 import { BookState } from '@interfaces/bookState';
 import { RootState } from '@interfaces/rootState';
-import { LibraryType } from '@interfaces/screens';
-import { LibraryStackParamList } from '@interfaces/navigatorParamLists';
 import actionCreators from '@redux/book/actions';
-import { trimLineBreaks } from '@utils/stringUtils';
 
 import styles from './styles';
```
Comment (sfernandez11) [inline wbooks/src/app/screens/Library/index.tsx:6]: Remove unused styles

File: wbooks/src/app/screens/SearchScreen/index.tsx [hunk 14]
```diff
@@ -0,0 +1,24 @@
...
+import useFilteredBooks from '@hooks/useFilteredBooks';
+import { BookState } from '@interfaces/bookState';
+import { RootState } from '@interfaces/rootState';
+
+import styles from './styles';
+
+function SearchScreen() {
+  const filteredBooks = useFilteredBooks();
+
+  const { searchedBook } = useSelector<RootState, BookState>(state => state.books);
```
Comment (sfernandez11) [inline wbooks/src/app/screens/SearchScreen/index.tsx]: Use the Hook to get this value => `const [searchedBook, filteredBooks] = useFilteredBooks();`

File: wbooks/src/app/screens/SearchScreen/index.tsx [hunk 15]
```diff
@@ -0,0 +1,24 @@
...
+
+  return (
+    <SafeAreaView style={styles.container}>
```
Comment (sfernandez11) [inline wbooks/src/app/screens/SearchScreen/index.tsx:13]: Same here, don't repeat `SafeAreaView`  in the same screen

File: wbooks/src/app/screens/SearchScreen/styles.ts [hunk 16]
```diff
@@ -0,0 +1,12 @@
+import { StyleSheet } from 'react-native';
+import colors from '@constants/colors';
+
+export default StyleSheet.create({
+  container: {
+    backgroundColor: colors.polar,
+    flex: 1
+  },
+  firstItem: {
```
Comment (sfernandez11) [inline wbooks/src/app/screens/SearchScreen/styles.ts]: Why??? Be careful with the copy/paste

**Review by sfernandez11 — CHANGES_REQUESTED** [review]
Great work Hernan! Some comments...

File: wbooks/src/app/components/AppNavigator/config.tsx (hunk 1 above)
Comment (hernan-clich) [reply wbooks/src/app/components/AppNavigator/config.tsx]: I needed a solution to have some common header elements that could be shared in order to obtain 2 different headers (with some similarities). After reading all your comments I realized the child screens will inherit them regardless.

File: wbooks/src/app/components/AppNavigator/config.tsx (hunk 3 above)
Comment (hernan-clich) [reply wbooks/src/app/components/AppNavigator/config.tsx]: Great to know, this one really is a game changer.

File: wbooks/src/app/components/AppNavigator/config.tsx (hunk 2 above)
Comment (hernan-clich) [reply wbooks/src/app/components/AppNavigator/config.tsx]: I'll include it in the mainHeader object.

File: wbooks/src/app/components/AppNavigator/index.tsx (hunk 4 above)
Comment (hernan-clich) [reply wbooks/src/app/components/AppNavigator/index.tsx]: Extracted

File: wbooks/src/app/components/AppNavigator/config.tsx (hunk 5 above)
Comment (hernan-clich) [reply wbooks/src/app/components/AppNavigator/config.tsx:30]: If I don't use that property, the back button will be displayed while it shouldn't (according to the Marvel designs).

I obtained this idea from [here](https://stackoverflow.com/a/43980393).

image

File: wbooks/src/app/components/AppNavigator/config.tsx (hunk 6 above)
Comment (hernan-clich) [reply wbooks/src/app/components/AppNavigator/config.tsx]: Refactored the entire AppNavigator/config.tsx file

File: wbooks/src/app/components/AppNavigator/index.tsx (hunk 7 above)
Comment (hernan-clich) [reply wbooks/src/app/components/AppNavigator/index.tsx]: I refactored so I only have to import named imports.

File: wbooks/src/app/components/AppNavigator/index.tsx (hunk 8 above)
Comment (hernan-clich) [reply wbooks/src/app/components/AppNavigator/index.tsx:20]: I'm no longer using and spreading the `sharedHeaderProperties`, so now the `mainHeader` has all the main properties, and the other headers only contain the customized features.

File: wbooks/src/app/components/SearchEmpty/index.tsx (hunk 9 above)
Comment (hernan-clich) [reply wbooks/src/app/components/SearchEmpty/index.tsx]: You're absolutely right.

File: wbooks/src/app/components/SearchBar/styles.ts (hunk 10 above)
Comment (hernan-clich) [reply wbooks/src/app/components/SearchBar/styles.ts]: Refactored

File: wbooks/src/app/components/SearchEmpty/styles.ts (hunk 11 above)
Comment (hernan-clich) [reply wbooks/src/app/components/SearchEmpty/styles.ts]: Refactored

File: wbooks/src/app/screens/Library/index.tsx (hunk 12 above)
Comment (hernan-clich) [reply wbooks/src/app/screens/Library/index.tsx:12]: I think it might be more beneficial to remove the `SafeAreaView` from the `BookList` component rather than the `Library` screen, so any potential new features in the `Library` screen would also be wrapped inside it.

File: wbooks/src/app/screens/Library/index.tsx (hunk 13 above)
Comment (hernan-clich) [reply wbooks/src/app/screens/Library/index.tsx:6]: Refactored

File: wbooks/src/app/screens/SearchScreen/index.tsx (hunk 15 above)
Comment (hernan-clich) [reply wbooks/src/app/screens/SearchScreen/index.tsx:13]: This one got settled by removing the `SafeAreaView` from the `BookList` component

File: wbooks/src/app/screens/SearchScreen/styles.ts (hunk 16 above)
Comment (hernan-clich) [reply wbooks/src/app/screens/SearchScreen/styles.ts]: Removed

File: wbooks/src/app/screens/SearchScreen/index.tsx (hunk 14 above)
Comment (hernan-clich) [reply wbooks/src/app/screens/SearchScreen/index.tsx]: I renamed the hook as `useBooks` and updated it so it returns `searchedBook`, `filteredBooks` and `books` (as in all books) to make it even more reusable, I kept the return type as an object, as it makes it easier to ignore values that are not needed.

**Review by sfernandez11 — APPROVED** [review]
🥇
